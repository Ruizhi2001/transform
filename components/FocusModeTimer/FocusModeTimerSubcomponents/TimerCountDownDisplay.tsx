import React, { useState } from "react";
import { Text, View, StyleSheet, Button } from "react-native";

type Props = {
  timerDate: Date;
};

export const TimerCountDownDisplay: React.FC<Props> = ({ timerDate }) => {
  return (
    <View>
      <Text style={styles.timerCountDownText}>
        {timerDate.getHours().toString().padStart(2, "0")}:
        {timerDate.getMinutes().toString().padStart(2, "0")}:
        {timerDate.getSeconds().toString().padStart(2, "0")}
      </Text>
    </View>
  );
};

const styles = StyleSheet.create({
  timerCountDownText: {
    fontSize: 90,
    fontWeight: "700",
  },
});
