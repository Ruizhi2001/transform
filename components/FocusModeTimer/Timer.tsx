import { StyleSheet, Text, View } from "react-native";
import React, { useEffect, useState } from "react";
import { TimerToggleButton } from "./FocusModeTimerSubcomponents/TimerToggleButton";
import { TimerCountDownDisplay } from "./FocusModeTimerSubcomponents/TimerCountDownDisplay";
import { TimeSelectorList } from "./FocusModeTimerSubcomponents/TimeSelectorList";
import {
  DocumentReference,
  doc,
  getDoc,
  increment,
  updateDoc,
} from "firebase/firestore";
import { auth, db } from "../../FirebaseConfig";
import { useAppDispatch } from "../../redux/hooks";
import { incrementByAmount } from "../../redux/expSlice";

const defaultDuration = 20 * 60 * 1000;

export const Timer: React.FC = () => {
  const [timerCount, setTimerCount] = useState<number>(defaultDuration);
  const [timerInterval, setTimerInterval] = useState<NodeJS.Timer | null>(null);
  const [isTimerRunning, setIsTimerRunning] = useState<boolean>(false);
  const user = auth.currentUser;

  let userDocRef: DocumentReference;
  const dispatch = useAppDispatch();

  if (user) {
    userDocRef = doc(db, "users", user.uid);
    getDoc(userDocRef);
  }

  const startTimer = () => {
    const id = setInterval(() => setTimerCount((prev) => prev - 1000), 1000);
    setIsTimerRunning(true);
    setTimerInterval(id);
  };

  const stopTimer = () => {
    setIsTimerRunning(false);
    if (timerInterval != null) {
      clearInterval(timerInterval);
      setTimerCount(defaultDuration);
    }
  };

  const setTimer = (duration: number) => {
    setTimerCount(duration * 60 * 1000);
  };

  useEffect(() => {
    if (timerCount === 0) {
      updateDoc(userDocRef, {
        exp: increment(timerCount),
      });
      dispatch(incrementByAmount(timerCount));
      stopTimer();
    }
  }, [timerCount]);

  return (
    <View style={styles.container}>
      <Text style={styles.timerText}>
        {isTimerRunning ? "TRANSFORMING!!!" : "TRANSFORM"}
      </Text>
      <TimerCountDownDisplay timerDate={new Date(timerCount)} />
      {isTimerRunning ? null : <TimeSelectorList toggleTimer={setTimer} />}
      <TimerToggleButton
        isTimerRunning={isTimerRunning}
        startTimer={startTimer}
        stopTimer={stopTimer}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    justifyContent: "space-between",
  },
  timerText: {
    fontSize: 30,
    alignItems: "center",
    fontWeight: "700",
  },
});
