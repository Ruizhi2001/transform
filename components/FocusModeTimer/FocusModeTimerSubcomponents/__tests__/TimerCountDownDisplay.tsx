import { defaultDuration } from "../../Timer";
import { TimerCountDownDisplay } from "../TimerCountDownDisplay";
import renderer from "react-test-renderer";

describe("TimerCountDownDisplay", () => {
  it("should match the snapshot", () => {
    const tree = renderer
      .create(<TimerCountDownDisplay timerDate={new Date(defaultDuration)} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
