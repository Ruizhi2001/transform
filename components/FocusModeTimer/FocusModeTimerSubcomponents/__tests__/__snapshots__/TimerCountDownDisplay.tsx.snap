// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimerCountDownDisplay should match the snapshot 1`] = `
<View>
  <Text
    style={
      {
        "fontSize": 90,
        "fontWeight": "700",
      }
    }
  >
    00
    :
    20
    :
    00
  </Text>
</View>
`;
